caddyserver/caddy
# Reuse unix sockets (UDS) and don't try to serve HTTP/3 over UDS

This PR hopefully fixes #5056 - any use of Unix domain sockets (UDS) cause errors now because somehow I missed that SO_REUSEPORT doesn't apply to UDS, even though it's a socket too.

My testing didn't catch this before because I didn't attempt to do a config reload with unix sockets specifically.

---

To test this, I created a simple server:

```
localhost

bind unix//home/matt/Desktop/test.sock
respond "Hello world!"
```

(I also enabled `debug` mode for my tests, but not required.)

I also added `time.Sleep(1 * time.Second)` to the `static_response` handler (`respond` directive) between writing headers and the body, to simulate a slow request-response. (I also repeated the tests without this sleep successfully.)

I started the server, then I ran this bash loop:

```
for i in {1..100000}; do
  curl -v --unix-socket ~/Desktop/test.sock "https://localhost/"
  sleep 0.001
done
```

to get "Hello world".

At the same time I ran this loop:

```
for i in {1..10000}; do
  curl -v "http://localhost:2019/load" -H "Content-Type: text/caddyfile" -H "Cache-Control: must-revalidate" --data-binary @Caddyfile
  sleep 0.01
done
```

to rapidly change the config. This should put the reload code (and unix socket reuse) through its paces.

I observed 0 errors, 0 failed requests, and 0 resource leaks (I used `localhost:2019/debug/pprof` to watch goroutines and I also used the system monitor to measure memory use).

---

For HTTP/3 support on UDS, we having an upstream discussion: https://github.com/lucas-clemente/quic-go/issues/3560

I don't expect the tests to pass for all platforms yet. I just found out Go has a new `unix` build tag that I'm trying out.
